client/js/widgets/widget.js

```diff
@@ -311,18 +354,80 @@ export class Widget extends StateManaged {
     const routine = this.get(property) !== undefined ? this.get(property) : property;
 
     for(const original of routine) {
-      const a = JSON.parse(JSON.stringify(original));
+      let a = JSON.parse(JSON.stringify(original));
+      if(typeof a == 'object')
+        a = evaluateVariablesRecursively(a);
       var problems = [];
 
       if(this.get('debug')) console.log(`${this.id}: ${JSON.stringify(original)}`);
 
-      if(a.applyVariables) this.applyVariables(a, variables, problems);
-
       if(a.skip) {
         $('#debugButtonOutput').textContent += '\n\n\nOPERATION SKIPPED: \n' + JSON.stringify(a, null, '  ');
         continue;
       }
 
+      if(typeof a == 'string') {
+        const identifier = '(?:[a-zA-Z0-9_-]|\\\\u[0-9a-fA-F]{4})+';
+        const string     = `'((?:[a-zA-Z0-9,.() _-]|\\\\u[0-9a-fA-F]{4})*)'`;
+        const number     = '(-?(?:0|[1-9][0-9]*)(?:\\.[0-9]+)?)';
+        const variable   = `(\\$\\{[^}]+\\})`;
+        const parameter  = `(null|true|false|\\[\\]|\\{\\}|${number}|${variable}|${string})`;
+
+        const left       = `var (\\$)?(${identifier})(?:\\.(\\$)?(${identifier}))?`;
+        const operation  = `${identifier}|[=+*/%<!>&|-]{1,2}`;
+
+        const regex      = `^${left} += +(?:${parameter}|(?:${parameter} +)?(${operation})(?: +${parameter})?(?: +${parameter})?(?: +${parameter})?)( +//.+)?`;
+
+        const match = a.match(new RegExp(regex + '\x24')); // the minifier doesn't like a "$" here
+
+        if(match) {
+          const getParam = (offset, defaultValue)=>{
+            if(typeof match[offset+3] == 'string') {
+              return unescape(match[offset+3]);
+            } else if(typeof match[offset+1] == 'string') {
+              return +match[offset+1];
+            } else if(match[offset] == '[]') {
+              return [];
+            } else if(match[offset] == '{}') {
+              return {};
+            } else if(match[offset] == 'null') {
+              return null;
+            } else if(match[offset] == 'true') {
+              return true;
+            } else if(match[offset] == 'false') {
+              return false;
+            } else if(match[offset] == 'false') {
+              return false;
+            } else if(typeof match[offset+2] == 'string') {
+              const result = evaluateVariables(match[offset+2]);
+              return result !== undefined ? result : defaultValue;
+            } else {
+              return 1;
+            }
+          };
+          const getValue = function(input) {
+            const toNum = s=>typeof s == 'string' && s.match(/^[-+]?[0-9]+(\.[0-9]+)?$/) ? +s : s;
```

Comment:
Should this use the number pattern defined earlier?

Reply:
In an ideal world, yes. This one has already been in the code for quite a while though and it turns numeric strings into numbers. I don't see a real reason to potentially break stuff by changing it (`"000"` would no longer be turned into `0`).

We could use this one in the expression parser but the other one should be more correct.